src/engine/renderer/gl_shader.cpp

```diff
@@ -1248,27 +1274,130 @@ void GLShaderManager::CompileGPUShaders( GLShader *shader, shaderProgram_t *prog
 		glAttachShader( program->program, _deformShaders[ deformIndex ] );
 	}
 	if ( shader->_hasFragmentShader ) {
-		glAttachShader( program->program, program->FS );
+		glAttachShader( program->program, baseProgram->FS );
 	}
 	if ( shader->_hasComputeShader ) {
-		glAttachShader( program->program, program->CS );
+		glAttachShader( program->program, baseProgram->CS );
 	}
 
 	BindAttribLocations( program->program );
 	LinkProgram( program->program );
 }
 
+// This will generate all the extra code for material system shaders
+std::string GLShaderManager::ShaderPostProcess( GLShader *shader, const std::string& shaderText ) {
+	std::string newShaderText;
+	std::string materialStruct = "\nstruct Material {\n";
+	std::string materialBlock = "layout(std430, binding = 0) readonly buffer materialsSSBO {\n"
+									 "  Material materials[];\n"
+									 "};\n\n";
+	std::string materialDefines;
+
+	/* Generate the struct and defines in the form of:
+	* struct Material {
+	*   type uniform0;
+	*   type uniform1;
+	*   ..
+	*   type uniformn;
+	* }
+	* 
+	* #define uniformx materials[baseInstance].uniformx
+	*/
+
+	for( GLUniform* uniform : shader->_uniforms ) {
+		if ( uniform->IsGlobal() ) {
+			continue;
+		}
+
+		if ( uniform->IsTexture() ) {
+			materialStruct += "  uvec2 ";
+			materialStruct += uniform->GetName();
+		} else {
+			materialStruct += "  " + uniform->GetType() + " " + uniform->GetName();
+		}
+
+		if ( uniform->GetComponentSize() ) {
+			materialStruct += "[ " + std::to_string( uniform->GetComponentSize() ) + " ]";
+		}
+		materialStruct += ";\n";
+
+		// vec3 is aligned to 4 components, so just pad it with int
+		// TODO: Try to move 1 component uniforms here to avoid wasting memory
+		if ( uniform->GetSTD430Size() == 3 ) {
+			materialStruct += "  int ";
+			materialStruct += uniform->GetName();
+			materialStruct += "_padding;\n";
+		}
+
+		materialDefines += "#define ";
+		materialDefines += uniform->GetName();
+
+		if ( uniform->IsTexture() ) { // Driver bug: AMD compiler crashes when using the SSBO value directly
+			materialDefines += "_initial uvec2("; // We'll need this to create sampler objects later
+		}
+
+		materialDefines += " materials[baseInstance].";
+		materialDefines += uniform->GetName();
+		
+		if ( uniform->IsTexture() ) {
+			materialDefines += " )";
+		}
+
+		materialDefines += "\n";
+	}
+
+	// Array of structs is aligned to the largest member of the struct
+	for ( uint i = 0; i < shader->padding; i++ ) {
+		materialStruct += "  int material_padding" + std::to_string( i );
+		materialStruct += ";\n";
+	}
+
+	materialStruct += "};\n\n";
+	materialDefines += "\n";
+
+	std::istringstream shaderTextStream( shaderText );
+	std::string shaderMain;
+
+	std::string line;
+	
+	while( std::getline( shaderTextStream, line, '\n' ) ) {
+		// Remove local uniform declarations, but avoid removing uniform/storage blocks
+		if( !( line.find( "uniform" ) == std::string::npos || line.find( ";" ) == std::string::npos ) ) {
```

Comment:
Most of the uniforms are replaced with buffer accesses in order to be able to draw everything with very few commands and state changes. So e. g. something like `u_Color` and `u_CurrentMap` will be added into the buffer and so has to be removed from the shader. Uniforms that are not unique to the surface, like u_ViewOrigin or u_ViewUp and u_ModelMatrix, u_ModelViewProjectionMatrix etc. are not removed since they're the same for each drawcall within the same view.

Reply:
I've added better explanation in comments in that part of the code.